File: backend/app/services/geo.py
import math
import re
from typing import Optional
from urllib.parse import quote

import httpx

from app.config import settings
from app.services.listing_fetcher import BROWSER_HEADERS

GEOCODE_CACHE: dict[str, Optional[tuple[float, float]]] = {}
DIRECTIONS_CACHE: dict[str, Optional[tuple[float, float, float]]] = {}

METERS_TO_MILES = 0.000621371
MAPBOX_MATRIX_MAX_COORDS = 25

COMMUTE_SPEED_MPH = {
    "walking": 3.0,
    "biking": 10.0,
    "transit": 18.0,
    "driving": 25.0,
}

MAPBOX_DIRECTIONS_PROFILE = {
    "walking": "walking",
    "driving": "driving",
    "biking": "cycling",
    "transit": "driving",
}


def haversine_miles(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    radius = 3958.8
    d_lat = math.radians(lat2 - lat1)
    d_lon = math.radians(lon2 - lon1)
    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(math.radians(lat1))
        * math.cos(math.radians(lat2))
        * math.sin(d_lon / 2) ** 2
    )
    return radius * 2 * math.asin(math.sqrt(a))


def estimate_commute_minutes(distance_miles: float, mode: str) -> int:
    speed = COMMUTE_SPEED_MPH.get(mode, COMMUTE_SPEED_MPH["walking"])
    if distance_miles <= 0:
        return 0
    return max(1, int(round(distance_miles / speed * 60)))


def max_commute_radius_miles(max_minutes: int, mode: str) -> float:
    speed = COMMUTE_SPEED_MPH.get(mode, COMMUTE_SPEED_MPH["walking"])
    return max_minutes / 60.0 * speed


def mapbox_directions_profile(commute_mode: str) -> str:
    return MAPBOX_DIRECTIONS_PROFILE.get(commute_mode, "walking")


def _commute_from_route(
    distance_meters: float,
    duration_seconds: float,
    commute_mode: str,
) -> tuple[float, int]:
    distance_miles = round(distance_meters * METERS_TO_MILES, 2)
    if commute_mode == "transit":
        minutes = estimate_commute_minutes(distance_miles, "transit")
    else:
        minutes = max(1, int(round(duration_seconds / 60)))
    return distance_miles, minutes


def _directions_cache_key(
    origin_lat: float,
    origin_lng: float,
    dest_lat: float,
    dest_lng: float,
    commute_mode: str,
) -> str:
    profile = mapbox_directions_profile(commute_mode)
    return (
        f"{profile}|{origin_lat:.5f},{origin_lng:.5f}|"
        f"{dest_lat:.5f},{dest_lng:.5f}"
    )


def mapbox_route_commute(
    origin_lat: float,
    origin_lng: float,
    dest_lat: float,
    dest_lng: float,
    commute_mode: str,
) -> Optional[tuple[float, int]]:
    """Route distance (mi) and travel time (min) via Mapbox Directions."""
    token = settings.mapbox_access_token
    if not token:
        return None

    cache_key = _directions_cache_key(
        origin_lat, origin_lng, dest_lat, dest_lng, commute_mode
    )
    if cache_key in DIRECTIONS_CACHE:
        cached = DIRECTIONS_CACHE[cache_key]
        if cached is None:
            return None
        distance_miles, minutes, _duration = cached
        return distance_miles, int(minutes)

    profile = mapbox_directions_profile(commute_mode)
    coordinates = f"{origin_lng},{origin_lat};{dest_lng},{dest_lat}"
    url = (
        "https://api.mapbox.com/directions/v5/mapbox/"
        f"{profile}/{coordinates}"
    )
    try:
        with httpx.Client(timeout=15, follow_redirects=True) as client:
            response = client.get(
                url,
                params={
                    "access_token": token,
                    "overview": "false",
                    "alternatives": "false",
                },
            )
            response.raise_for_status()
            routes = response.json().get("routes") or []
            if not routes:
                DIRECTIONS_CACHE[cache_key] = None
                return None
            route = routes[0]
            distance_meters = float(route["distance"])
            duration_seconds = float(route["duration"])
            distance_miles, minutes = _commute_from_route(
                distance_meters, duration_seconds, commute_mode
            )
            DIRECTIONS_CACHE[cache_key] = (
                distance_miles,
                float(minutes),
                duration_seconds,
            )
            return distance_miles, minutes
    except Exception:
        DIRECTIONS_CACHE[cache_key] = None
        return None


def mapbox_batch_route_commutes(
    campus_lat: float,
    campus_lng: float,
    destinations: list[tuple[int, float, float]],
    commute_mode: str,
) -> dict[int, tuple[float, int]]:
    """Batch route commutes from campus to many listings via Mapbox Matrix."""
    token = settings.mapbox_access_token
    if not token or not destinations:
        return {}

    profile = mapbox_directions_profile(commute_mode)
    results: dict[int, tuple[float, int]] = {}
    chunk_size = MAPBOX_MATRIX_MAX_COORDS - 1

    for chunk_start in range(0, len(destinations), chunk_size):
        chunk = destinations[chunk_start : chunk_start + chunk_size]
        coords = [(campus_lat, campus_lng)] + [(lat, lng) for _, lat, lng in chunk]
        coord_str = ";".join(f"{lng},{lat}" for lat, lng in coords)
        destination_indices = ";".join(str(index) for index in range(1, len(coords)))

        url = (
            "https://api.mapbox.com/directions-matrix/v1/mapbox/"
            f"{profile}/{coord_str}"
        )
        try:
            with httpx.Client(timeout=20, follow_redirects=True) as client:
                response = client.get(
                    url,
                    params={
                        "access_token": token,
                        "sources": "0",
                        "destinations": destination_indices,
                        "annotations": "distance,duration",
                    },
                )
                response.raise_for_status()
                payload = response.json()
                distances = payload.get("distances") or []
                durations = payload.get("durations") or []
                if not distances or not durations:
                    continue

                row_distances = distances[0]
                row_durations = durations[0]
                for index, (listing_id, _, _) in enumerate(chunk):
                    if index >= len(row_distances) or index >= len(row_durations):
                        continue
                    distance_meters = row_distances[index]
                    duration_seconds = row_durations[index]
                    if distance_meters is None or duration_seconds is None:
                        continue
                    results[listing_id] = _commute_from_route(
                        float(distance_meters),
                        float(duration_seconds),
                        commute_mode,
                    )
        except Exception:
            continue

    return results


def commute_between_coords(
    origin_lat: float,
    origin_lng: float,
    dest_lat: float,
    dest_lng: float,
    commute_mode: str,
) -> Optional[tuple[float, int]]:
    """Prefer Mapbox route distance/time; fall back to straight-line estimate."""
    routed = mapbox_route_commute(
        origin_lat, origin_lng, dest_lat, dest_lng, commute_mode
    )
    if routed is not None:
        return routed

    distance = haversine_miles(origin_lat, origin_lng, dest_lat, dest_lng)
    minutes = estimate_commute_minutes(distance, commute_mode)
    return round(distance, 2), minutes


def _cache_key(query: str) -> str:
    return re.sub(r"\s+", " ", query.strip().lower())


def geocode(query: str) -> Optional[tuple[float, float]]:
    """Resolve an address or place name to (lat, lng)."""
    key = _cache_key(query)
    if not key:
        return None
    if key in GEOCODE_CACHE:
        return GEOCODE_CACHE[key]

    coords: Optional[tuple[float, float]] = None
    if settings.mapbox_access_token:
        coords = _geocode_mapbox(query)
    if coords is None:
        coords = _geocode_nominatim(query)

    GEOCODE_CACHE[key] = coords
    return coords


def _geocode_mapbox(query: str) -> Optional[tuple[float, float]]:
    token = settings.mapbox_access_token
    if not token:
        return None
    url = (
        "https://api.mapbox.com/geocoding/v5/mapbox.places/"
        f"{quote(query)}.json"
    )
    try:
        with httpx.Client(timeout=12, follow_redirects=True) as client:
            response = client.get(
                url,
                params={
                    "access_token": token,
                    "country": "us",
                    "limit": 1,
                },
            )
            response.raise_for_status()
            features = response.json().get("features") or []
            if not features:
                return None
            lng, lat = features[0]["center"]
            return float(lat), float(lng)
    except Exception:
        return None


def _geocode_nominatim(query: str) -> Optional[tuple[float, float]]:
    headers = {
        **BROWSER_HEADERS,
        "User-Agent": "NestMatchAI/1.0 (campus apartment search)",
    }
    try:
        with httpx.Client(headers=headers, timeout=12, follow_redirects=True) as client:
            response = client.get(
                "https://nominatim.openstreetmap.org/search",
                params={
                    "q": query,
                    "format": "json",
                    "limit": 1,
                    "countrycodes": "us",
                },
            )
            response.raise_for_status()
            results = response.json()
            if not results:
                return None
            return float(results[0]["lat"]), float(results[0]["lon"])
    except Exception:
        return None

File: frontend/src/lib/geo.ts
/** Shared geocoding and Mapbox route helpers for commute estimates. */

import { fetchAppConfig } from '@/lib/api'
import type { CommuteMode } from '@/types/studentProfile'

/** Fallback speeds when Mapbox is unavailable (match backend geo.py). */
export const COMMUTE_SPEED_MPH: Record<CommuteMode, number> = {
  walking: 3,
  biking: 10,
  transit: 18,
  driving: 25,
}

const MAPBOX_DIRECTIONS_PROFILE: Record<CommuteMode, string> = {
  walking: 'walking',
  driving: 'driving',
  biking: 'cycling',
  transit: 'driving',
}

const METERS_TO_MILES = 0.000621371

const geocodeCache = new Map<string, Promise<[number, number] | null>>()
const routeCache = new Map<
  string,
  Promise<{ distanceMiles: number; minutes: number } | null>
>()

export function haversineMiles(
  a: [number, number],
  b: [number, number],
): number {
  const [lon1, lat1] = a
  const [lon2, lat2] = b
  const radius = 3958.8
  const dLat = ((lat2 - lat1) * Math.PI) / 180
  const dLon = ((lon2 - lon1) * Math.PI) / 180
  const x =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) ** 2
  return radius * 2 * Math.asin(Math.sqrt(x))
}

export async function geocodeAddress(
  query: string,
  token: string,
  proximity?: [number, number],
): Promise<[number, number] | null> {
  const normalized = query.trim().toLowerCase()
  if (!normalized) return null

  const cacheKey = `${normalized}|${proximity?.join(',') ?? ''}`
  const cached = geocodeCache.get(cacheKey)
  if (cached) return cached

  const promise = (async () => {
    const params = new URLSearchParams({
      access_token: token,
      limit: '1',
      country: 'us',
    })
    if (proximity) {
      params.set('proximity', `${proximity[0]},${proximity[1]}`)
    }
    const url = `https://api.mapbox.com/geocoding/v5/mapbox.places/${encodeURIComponent(query)}.json?${params}`
    const response = await fetch(url)
    if (!response.ok) return null
    const data = (await response.json()) as {
      features?: { center: [number, number] }[]
    }
    return data.features?.[0]?.center ?? null
  })()

  geocodeCache.set(cacheKey, promise)
  return promise
}

export function estimateCommuteMinutes(
  distanceMiles: number,
  mode: CommuteMode,
): number {
  if (distanceMiles <= 0) return 0
  const speed = COMMUTE_SPEED_MPH[mode] ?? COMMUTE_SPEED_MPH.walking
  return Math.max(1, Math.round((distanceMiles / speed) * 60))
}

function commuteFromRoute(
  distanceMeters: number,
  durationSeconds: number,
  mode: CommuteMode,
): { distanceMiles: number; minutes: number } {
  const distanceMiles = Math.round(distanceMeters * METERS_TO_MILES * 100) / 100
  const minutes =
    mode === 'transit'
      ? estimateCommuteMinutes(distanceMiles, 'transit')
      : Math.max(1, Math.round(durationSeconds / 60))
  return { distanceMiles, minutes }
}

/** Mapbox Directions route distance (mi) and travel time (min) between two points. */
export async function fetchMapboxRouteCommute(
  origin: [number, number],
  destination: [number, number],
  token: string,
  mode: CommuteMode,
): Promise<{ distanceMiles: number; minutes: number } | null> {
  const profile = MAPBOX_DIRECTIONS_PROFILE[mode]
  const cacheKey = `${profile}|${origin.join(',')}|${destination.join(',')}`
  const cached = routeCache.get(cacheKey)
  if (cached) return cached

  const promise = (async () => {
    const coordinates = `${origin[0]},${origin[1]};${destination[0]},${destination[1]}`
    const params = new URLSearchParams({
      access_token: token,
      overview: 'false',
      alternatives: 'false',
    })
    const url = `https://api.mapbox.com/directions/v5/mapbox/${profile}/${coordinates}?${params}`
    const response = await fetch(url)
    if (!response.ok) return null

    const data = (await response.json()) as {
      routes?: { distance: number; duration: number }[]
    }
    const route = data.routes?.[0]
    if (!route) return null

    return commuteFromRoute(route.distance, route.duration, mode)
  })()

  routeCache.set(cacheKey, promise)
  return promise
}

/** Geocode campus + listing and return Mapbox route miles and minutes. */
export async function fetchMapboxCommuteBetweenAddresses(
  campusAddress: string,
  listingAddress: string,
  mode: CommuteMode,
): Promise<{ distanceMiles: number; minutes: number } | null> {
  const config = await fetchAppConfig()
  if (!config.mapboxToken) return null

  const campusCoords = await geocodeAddress(campusAddress, config.mapboxToken)
  if (!campusCoords) return null

  const listingCoords = await geocodeAddress(
    listingAddress,
    config.mapboxToken,
    campusCoords,
  )
  if (!listingCoords) return null

  const routed = await fetchMapboxRouteCommute(
    campusCoords,
    listingCoords,
    config.mapboxToken,
    mode,
  )
  if (routed) return routed

  const distanceMiles =
    Math.round(haversineMiles(campusCoords, listingCoords) * 100) / 100
  return {
    distanceMiles,
    minutes: estimateCommuteMinutes(distanceMiles, mode),
  }
}

/** @deprecated use fetchMapboxCommuteBetweenAddresses */
export async function fetchDistanceMilesBetweenAddresses(
  campusAddress: string,
  listingAddress: string,
): Promise<number | null> {
  const estimate = await fetchMapboxCommuteBetweenAddresses(
    campusAddress,
    listingAddress,
    'walking',
  )
  return estimate?.distanceMiles ?? null
}

export async function estimateCommuteClientSide(
  campusAddress: string,
  listingAddress: string,
  mode: CommuteMode,
): Promise<{ minutes: number; distanceMiles: number } | null> {
  return fetchMapboxCommuteBetweenAddresses(campusAddress, listingAddress, mode)
}
